modules/core/src/main/java/org/apache/ignite/internal/processors/localtask/DurableBackgroundTasksProcessor.java

```diff
@@ -16,315 +16,342 @@
+                TASK_PREFIX,
+                (k, v) -> {
+                    DurableBackgroundTask t = (DurableBackgroundTask)v;
+
+                    tasks.put(t.name(), new DurableBackgroundTaskState(t, null, true));
+                },
+                true
+            );
+        });
+    }
 
-                super.cancel();
-            }
+    /** {@inheritDoc} */
+    @Override public void onReadyForReadWrite(ReadWriteMetastorage metastorage) {
+        ((GridCacheDatabaseSharedManager)ctx.cache().context().database()).addCheckpointListener(this);
+    }
 
-            @Override protected void body() {
-                try {
-                    if (forbidStartingNewTasks)
-                        return;
+    /** {@inheritDoc} */
+    @Override public void beforeCheckpointBegin(Context ctx) {
+        /* No op. */
+    }
 
-                    log.info("Executing durable background task: " + task.shortName());
+    /** {@inheritDoc} */
+    @Override public void onMarkCheckpointBegin(Context ctx) {
+        for (Iterator<Entry<String, DurableBackgroundTaskState>> it = tasks.entrySet().iterator(); it.hasNext(); ) {
+            DurableBackgroundTaskState taskState = it.next().getValue();
 
-                    task.execute(ctx);
+            if (taskState.state() == COMPLETED) {
+                assert taskState.saved();
 
-                    task.complete();
+                DurableBackgroundTask t = taskState.task();
 
-                    log.info("Execution of durable background task completed: " + task.shortName());
-                }
-                catch (Throwable e) {
-                    log.error("Could not execute durable background task: " + task.shortName(), e);
-                }
-                finally {
-                    startedTasks.remove(task.shortName());
+                toRmv.put(t.name(), t);
 
-                    asyncDurableBackgroundTaskWorkers.remove(this);
-                }
+                it.remove();
             }
-        };
-
-        asyncDurableBackgroundTaskWorkers.add(worker);
-
-        Thread asyncTask = new IgniteThread(worker);
-
-        asyncTask.start();
+        }
     }
 
     /** {@inheritDoc} */
-    @Override public void onKernalStart(boolean active) {
-        asyncDurableBackgroundTasksExecution();
+    @Override public void onCheckpointBegin(Context ctx) {
+        /* No op. */
     }
 
     /** {@inheritDoc} */
-    @Override public void onKernalStop(boolean cancel) {
-        forbidStartingNewTasks = true;
+    @Override public void afterCheckpointEnd(Context ctx) {
+        for (Iterator<Entry<String, DurableBackgroundTask>> it = toRmv.entrySet().iterator(); it.hasNext(); ) {
+            DurableBackgroundTask t = it.next().getValue();
 
-        awaitForWorkersStop(asyncDurableBackgroundTaskWorkers, true, log);
-    }
+            metaStorageOperation(metaStorage -> {
+                if (metaStorage != null && toRmv.containsKey(t.name()))
```

Comment:
What's the point of **toRmv.containsKey(t.name())** check?

Reply:
We discussed it in a personal.